ethcore/src/engines/mod.rs

```diff
@@ -451,6 +463,31 @@ pub trait EthEngine: Engine<::machine::EthereumMachine> {
 		self.machine().verify_transaction_basic(t, header)
 	}
 
+	/// Prepare the environment information passed for transaction execution.
+	fn prepare_env_info(&self, t: &SignedTransaction, block: &ExecutedBlock, env_info: &mut EnvInfo) {
```

Comment:
seems like all these new functions could be replaced with a more generic one on the main engine trait `fn execution_rules(block: Block) -> Machine::ExecutionRules`. For EthereumMachine this could be an object that provides the ordering rules and transaction outcome.

Reply:
@rphmeier Can you explain how you vision this to work? We don't need transaction ordering any more so there'll be two more functions -- `prepare_env_info` and `verify_transaction_outcome`. Both of them are actually really Casper-specific (it sets a block metadata value for `vote_gas_used`). Is the intention to create `ExecutionRules` struct to reduce size of `EthereumMachine`? Or would it be better if we just do https://github.com/paritytech/parity/pull/8654#discussion_r189244580 and delegate all Casper-related stuff there?

Reply:
we may want to use the `Engine` trait in deployments where there are no transactions. Casper may even work in that case because it is sufficiently general. Leaking the `SignedTransaction` and `ExecutedBlock` type into the engine trait makes all prior generalization work irrelevant.

Reply:
@rphmeier Currently the newly defined two functions `prepare_env_info` and `verify_transaction_outcome` is in the `EthEngine` trait. We didn't change `Engine` trait, nor leaked any additional types to `Engine`. So I think it would be okay from this point of view? I still don't have a good idea what the `Machine::ExecutionRules` type would look like.

And regarding Casper being general -- I think it may work in some future versions of Casper, but right now, the Hybrid Casper version is really dependent on executing transactions. The issue is that `simple_casper` being a contract, for many times it "bypass" engine -- functions like `deposit` and `vote` are just normal (signed or unsigned) transactions, so in some sense it's kind of "invisible" to engine.